cmd/gateway-b2.go

```diff
@@ -0,0 +1,650 @@
+	if err != nil {
+		return nil, err
+	}
+	return &b2Objects{
+		b2Client: client,
+		anonClient: &http.Client{
+			Transport: newCustomHTTPTransport(),
+		},
+		ctx: ctx,
+	}, nil
+}
+
+// Convert B2 errors to minio object layer errors.
+func b2ToObjectError(err error, params ...string) error {
+	if err == nil {
+		return nil
+	}
+
+	e, ok := err.(*Error)
+	if !ok {
+		// Code should be fixed if this function is called without doing traceError()
+		// Else handling different situations in this function makes this function complicated.
+		errorIf(err, "Expected type *Error")
+		return err
+	}
+
+	err = e.e
+	bucket := ""
+	object := ""
+	uploadID := ""
+	if len(params) >= 1 {
+		bucket = params[0]
+	}
+	if len(params) == 2 {
+		object = params[1]
+	}
+	if len(params) == 3 {
+		uploadID = params[2]
+	}
+
+	// Following code is a non-exhaustive list.
+	// for a more complete information - https://www.backblaze.com/b2/docs/
+	statusCode, code, msg := b2.Code(err)
+	if statusCode == 0 {
+		// We don't interpret non B2 errors. As B2 errors will
+		// have StatusCode to help to convert to object errors.
+		return e
+	}
+
+	switch code {
+	case "duplicate_bucket_name":
+		err = BucketAlreadyOwnedByYou{Bucket: bucket}
+	case "bad_request":
+		if object != "" {
+			err = ObjectNameInvalid{bucket, object}
+		} else if bucket != "" {
+			err = BucketNotFound{Bucket: bucket}
+		}
+	case "bad_bucket_id":
+		err = BucketNotFound{Bucket: bucket}
+	case "file_not_present", "not_found":
+		err = ObjectNotFound{bucket, object}
+	case "cannot_delete_non_empty_bucket":
+		err = BucketNotEmpty{bucket, ""}
+	}
+	if strings.Contains(msg, "No active upload for") && uploadID != "" {
+		err = InvalidUploadID{uploadID}
+	}
+
+	e.e = err
+	return e
+}
+
+// Shutdown saves any gateway metadata to disk
+// if necessary and reload upon next restart.
+func (l *b2Objects) Shutdown() error {
+	// TODO
+	return nil
+}
+
+// StorageInfo is not relevant to B2 backend.
+func (l *b2Objects) StorageInfo() (si StorageInfo) {
+	return si
+}
+
+// MakeBucket creates a new container on B2 backend.
+func (l *b2Objects) MakeBucketWithLocation(bucket, location string) error {
+	info := make(map[string]string)
+	info["x-minio-bucket-createTime"] = UTCNow().Format(time.RFC1123)
+	// All buckets are set to private by default.
+	_, err := l.b2Client.CreateBucket(l.ctx, bucket, "allPrivate", info, nil)
+	return b2ToObjectError(traceError(err), bucket)
+}
+
+// Bucket - bucket
+func (l *b2Objects) Bucket(bucket string) (*b2.Bucket, error) {
+	bktList, err := l.b2Client.ListBuckets(l.ctx)
+	if err != nil {
+		return nil, b2ToObjectError(traceError(err), bucket)
+	}
+	for _, bkt := range bktList {
+		if bkt.Name == bucket {
+			return bkt, nil
+		}
+	}
+	return nil, traceError(BucketNotFound{Bucket: bucket})
+}
+
+// Looks for x-minio-bucket-createTime as part of bucket metadata, if key
+// is absent then returns current UTC time instead. Returns error if
```

Comment:
> time.Unix(0,0) is not a good idea because of the tests that expect CreatTime to be always within some acceptable boundary of time

Which tests are you referring to? If it is unit tests we can (and should) change it as fit. This suggestion is to avoid any confusion to the user, who knows that the bucket was created much before the time we might show (i.e, last accessed time like you mention above).

Reply:
I am talking about mint tests.. Setting time.Unix(0, 0) will be more confusing to user IMO. I would prefer not make this change and keep it the way it is. We can even discuss about this but i would like to avoid epoch date.

For the most part users using buckets from minio gateway will not have any problems, B2 users aren't expecting this value anyways. So i choose to set this value to something more meaningful than Epoch time.

Reply:
As a matter of fact perhaps i can remove the x-minio-bucket-createTime and avoid such intelligence in the first place.

Reply:
> Setting time.Unix(0, 0) will be more confusing to user IMO. I would prefer not make this 

The idea behind `time.Unix(0,0)` is that it is a constant and doesn't change with time like `time.Now()`. We could choose a different constant `time.Time` value if that's the only issue.

> As a matter of fact perhaps i can remove the x-minio-bucket-createTime and avoid such intelligence in the first place.

what would `InitiatedTime` value for `BucketInfo` be?

Reply:
Nothing just time.Unix(0, 0) there is no such info available from B2 API so we don't return anything.